crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,496 @@
+    // always know the executed amount in case of partial fill.
+    pub fn total_fees(
+        &self,
+        external_prices: &ExternalPrices,
+        orders: &[Order],
+        configuration: &FeeConfiguration,
+    ) -> U256 {
+        self.trades.iter().fold(0.into(), |acc, trade| {
+            match orders
+                .iter()
+                .find(|order| order.signature == trade.signature.0)
+            {
+                Some(order) => {
+                    acc + match fee(trade, external_prices, order, configuration) {
+                        Some(fee) => fee,
+                        None => {
+                            tracing::warn!("possible incomplete fee calculation");
+                            0.into()
+                        }
+                    }
+                }
+                None => {
+                    tracing::warn!(
+                        "order not found for trade, possible incomplete fee calculation"
+                    );
+                    acc
+                }
+            }
+        })
+    }
+}
+
+fn surplus(
+    trade: &DecodedTrade,
+    tokens: &[Address],
+    clearing_prices: &[U256],
+    external_prices: &ExternalPrices,
+) -> Option<U256> {
+    let sell_token_index = trade.sell_token_index.as_u64() as usize;
+    let buy_token_index = trade.buy_token_index.as_u64() as usize;
+
+    let sell_token_clearing_price = clearing_prices.get(sell_token_index)?.to_big_rational();
+    let buy_token_clearing_price = clearing_prices.get(buy_token_index)?.to_big_rational();
+    let kind = order_kind(&trade.flags).unwrap();
+
+    if match kind {
+        OrderKind::Sell => &buy_token_clearing_price,
+        OrderKind::Buy => &sell_token_clearing_price,
+    }
+    .is_zero()
+    {
+        return None;
+    }
+
+    let surplus = trade_surplus(
+        kind,
+        &trade.sell_amount.to_big_rational(),
+        &trade.buy_amount.to_big_rational(),
+        &trade.executed_amount.to_big_rational(),
+        &sell_token_clearing_price,
+        &buy_token_clearing_price,
+    )?;
+
+    let normalized_surplus = match kind {
+        OrderKind::Sell => {
+            let buy_token = tokens.get(buy_token_index)?;
+            external_prices.try_get_native_amount(*buy_token, surplus / buy_token_clearing_price)?
+        }
+        OrderKind::Buy => {
+            let sell_token = tokens.get(sell_token_index)?;
+            external_prices
+                .try_get_native_amount(*sell_token, surplus / sell_token_clearing_price)?
+        }
+    };
+
+    big_rational_to_u256(&normalized_surplus).ok()
+}
+
+fn fee(
+    trade: &DecodedTrade,
+    external_prices: &ExternalPrices,
+    order: &Order,
+    configuration: &FeeConfiguration,
+) -> Option<U256> {
+    let full_fee_amount = u256_to_big_rational(&order.full_fee_amount);
+    let scaled_fee_amount = full_fee_amount * configuration.fee_objective_scaling_factor.clone();
```

Comment:
I think we already have a separate field in `Order` that contains the scaled unsubsidized fee amount that should be used for observed value computation. We should use that.

Reply:
cc @MartinquaXD as this is something you worked on if I remember correctly.

Reply:
Yes, this was part of #1199. The field that solvers should use for their objective value computation / fee calculation is `solver_fee`.

Reply:
Ah OK - nevermind then.